draft-ietf-quic-recovery.md

```diff
@@ -1580,14 +1586,18 @@ Invoked when an ACK frame with an ECN section is received from the peer.
 Invoked when DetectAndRemoveLostPackets deems packets lost.
 
 ~~~
-   InPersistentCongestion(lost_packets):
+   InPersistentCongestion(largest_lost):
+     // Persistent congestion cannot be declared on the
+     // first RTT sample.
+     if (is first RTT sample):
+       return false
```

Comment:
I'm not sure if that's correct. What matters is that the packet you use for the start of the persistent congestion period was sent after you already had an RTT sample.

Reply:
I agree with @marten-seemann here.

Consider the case where an endpoint receives the first ACK 10 seconds after sending the first packet (with an RTT of 10ms), then receives the next ACK in 20ms (with RTT of 10ms).

When the second ACK is being processed, the execution would pass through this `if`, and invoke `AreAllPacketsLost`. Because all the packets that were sent during the first few seconds are not acked, persistent congestion would be declared.

I think we can simply remove these lines, and rely on the fact that the comment in the pseudo-code talking about "edges." We can clarify what "edge" means, if necessary.

Reply:
> I think we can simply remove these lines, and rely on the fact that the comment in the pseudo-code talking about "edges." We can clarify what "edge" means, if necessary.

I'm not sure if that's sufficient. I think I'd prefer to make the pseudo-code more closely resemble what you'd have to implement. Specifically, to implement this, you'll need a new global variable `time_of_first_rtt_measurement`, or alternative `first_packet_sent_with_measured_rtt`, and then you'll have to implement a comparison with the time of the packets in `lost_packets`.

Reply:
> Specifically, to implement this, you'll need a new global variable `time_of_first_rtt_measurement`, or alternative `first_packet_sent_with_measured_rtt`, and then you'll have to implement a comparison with the time of the packets in `lost_packets`.

Can you implement that way? Persistent congestion is declared across all packet number spaces. That means that you'd have to consult if there any acks were received in the loss period observed in `lost_packets`. But you cannot remember all the moments when acks were received, because the state would be unbounded.

Reply:
@kazuho can you explain your reasoning around "edges" being sufficient?

Reply:
@mjoras If we define "edge" as a lost packet next to a packet that has been acked, having both edges means that a series of packets being lost are surrounded by acks. As there would have been RTT samples obtained for each of the surrounding acks, there is no need for a separate condition that checks if an RTT sample has been previously obtained.

Reply:
That's very cute. I would be okay with that too if we cleaned up what "edge" means and also clarify the notion of which contiguous segment should be considered for the persistent period.

Reply:
Discussed offline with @janaiyengar, it has turned out that the approach of tweaking the definion of edges does not work well, because the edges (from current time) can be any packet, while RTT can only be established with an ack-eliciting packet.